src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -288,112 +319,46 @@ HeapWord* ShenandoahFreeSet::try_allocate_in(ShenandoahHeapRegion* r, Shenandoah
-      if ((free != usable_free) && (free - usable_free < ShenandoahHeap::min_fill_size() * HeapWordSize)) {
-        // We'll have to add another card's memory to the padding
-        if (usable_free >= CardTable::card_size()) {
-          usable_free -= CardTable::card_size();
-        } else {
-          assert(usable_free == 0, "usable_free is a multiple of card_size and card_size > min_fill_size");
-        }
-      }
-      free /= HeapWordSize;
-      usable_free /= HeapWordSize;
-      size_t remnant = size % CardTable::card_size_in_words();
-      if (remnant > 0) {
-        // Since we have Elastic TLABs, align size up.  This is consistent with aligning min_size up.
-        size = size - remnant + CardTable::card_size_in_words();
-      }
+      // Since we have Elastic TLABs, align size up.  This is consistent with aligning min_size up.
+      size_t usable_free = get_usable_free_words(r->free());
+      size_t size = align_up(req.size(), CardTable::card_size_in_words());
+      size_t adjusted_min_size = align_up(req.min_size(), CardTable::card_size_in_words());
+
       if (size > usable_free) {
         size = usable_free;
-        assert(size % CardTable::card_size_in_words() == 0, "usable_free is a multiple of card table size");
       }
 
-      size_t adjusted_min_size = req.min_size();
-      remnant = adjusted_min_size % CardTable::card_size_in_words();
-      if (remnant > 0) {
-        // Round up adjusted_min_size to a multiple of alignment size
-        adjusted_min_size = adjusted_min_size - remnant + CardTable::card_size_in_words();
-      }
       if (size >= adjusted_min_size) {
-        result = r->allocate_aligned(size, req, CardTable::card_size());
-        assert(result != nullptr, "Allocation cannot fail");
-        size = req.actual_size();
-        assert(r->top() <= r->end(), "Allocation cannot span end of region");
-        // actual_size() will be set to size below.
-        assert((result == nullptr) || (size % CardTable::card_size_in_words() == 0),
-               "PLAB size must be multiple of card size");
-        assert((result == nullptr) || (((uintptr_t) result) % CardTable::card_size_in_words() == 0),
-               "PLAB start must align with card boundary");
-        if (free > usable_free) {
-          // Account for the alignment padding
-          size_t padding = (free - usable_free) * HeapWordSize;
-          increase_used(padding);
-          assert(r->is_old(), "All PLABs reside in old-gen");
-          _heap->old_generation()->increase_used(padding);
-          // For verification consistency, we need to report this padding to _heap
-          _heap->increase_used(padding);
-        }
+        result = allocated_aligned_plab(size, req, r);
       }
       // Otherwise, leave result == nullptr because the adjusted size is smaller than min size.
     } else {
       // This is a GCLAB or a TLAB allocation
+      size_t size = req.size();
```

Comment:
I prefer `adjusted_size` here too, but ok if you want to leave it this way (there are adjustments that happen below).